File: Charts/inflation_scenarios.py
# Export fixed VAR coefficients and oil paths for the browser scenario editor.
# =============================================================================
#%% Imports
import importlib.util
import json
import sys
from functools import lru_cache
from html import escape

import numpy as np
import pandas as pd
import plotly.io as pio
from statsmodels.tsa.api import VAR

import results_dashboard_config as config

DRAWABLE_DRIVERS = {
    "oil": dict(label="Oil price", unit="USD per barrel", minimum=0.01, maximum=None, keyStep=1),
    "unemployment": dict(label="Unemployment rate", unit="Per cent", minimum=0, maximum=100, keyStep=0.1),
    "expectations": dict(label="Inflation expectations", unit="Per cent", minimum=None, maximum=None, keyStep=0.1),
}


# =============================================================================
#%% Reuse the country workbook readers and monthly model specifications
def load_module(name, path):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


@lru_cache(maxsize=None)
def country_inputs(country):
    folder = config.ROOT / country / "Inflation Model"
    data = load_module(f"scenario_data_{country}", folder / "data.py")
    if country == "AU":
        quarterly = data.load_sheet("AU", "Q")
        monthly = data.load_sheet("AU-M", "M")
        drivers = quarterly[["Unemployment Rate", "Inflation expectation", "Import prices"]].copy()
        drivers.columns = ["unemployment", "expectations", "import_prices"]
        for name, column in [("oil", "Oil"), ("stage2", "Stage2")]:
            values = monthly[column].dropna()
            drivers[name] = values.groupby(values.index.asfreq("Q")).mean()
        drivers = drivers.loc["2001Q1":]
        targets = {name: quarterly[column].dropna().loc["2001Q1":].rename("cpi")
                   for name, column in {"headline": "Headline - Index", "trimmed_mean": "Trimmed mean index",
                                        "ex_food_energy": "Ex food & energy Index s.a"}.items()}
        return drivers, targets, ["cpi", "import_prices", "stage2"], 6
    previous = {name: sys.modules.get(name) for name in ("data", "varmodel")}
    try:
        sys.modules["data"] = data
        sys.modules["varmodel"] = load_module(f"scenario_var_{country}", folder / "varmodel.py")
        model = load_module(f"scenario_{country}", folder / f"{country.lower()}.py")
    finally:
        for name, module in previous.items():
            if module is None:
                sys.modules.pop(name, None)
            else:
                sys.modules[name] = module
    measures = (model.NATIONAL | model.TOKYO) if country == "JP" else (model.CPI_MEASURES | model.PPI_MEASURES)
    drivers = data.build_frame(model.DRIVERS, freq="M", start=model.START)
    targets = {name: data.series("JN" if country == "JP" else "KR", "M", column, name="cpi").loc[model.START:]
               for name, column in measures.items()}
    changes = ["cpi", "import_prices"] + (["stage2"] if country == "JP" else [])
    return drivers, targets, changes, 12


# =============================================================================
#%% Fit exactly the existing specifications and check against saved forecasts
def scenario_payload(country, spec):
    drivers, targets, changes, maxlags = country_inputs(country)
    saved = pd.read_csv(spec["path"], index_col=0)
    saved.index = pd.PeriodIndex(saved.index, freq=spec["frequency"])
    models = []
    for column, label in spec["series"].items():
        name = column.removesuffix("_yoy")
        target = targets[name]
        frame = pd.concat([target, drivers], axis=1).dropna()
        frame[changes] = 100 * frame[changes].pct_change(fill_method=None)
        frame = frame.dropna()
        expected = pd.period_range(frame.index[0], frame.index[-1], freq=spec["frequency"])
        if not frame.index.equals(expected):
            raise ValueError(f"Non-contiguous VAR sample: {country} {name}")
        fit = VAR(frame.to_numpy()).fit(maxlags=maxlags, ic="aic")
        steps = spec["forecast_steps"]
        baseline = (fit.forecast(frame.to_numpy()[-fit.k_ar:], steps) if fit.k_ar
                    else np.tile(fit.intercept, (steps, 1)))
        dates = pd.period_range(frame.index[-1] + 1, periods=steps, freq=spec["frequency"])
        year = 4 if spec["frequency"] == "Q" else 12
        history = target.loc[:frame.index[-1]].dropna()
        levels = pd.concat([history, pd.Series(history.iloc[-1] * np.cumprod(1 + baseline[:, 0] / 100), index=dates)])
        yoy = 100 * (levels / levels.shift(year) - 1)
        # Never attach new coefficients to a stale or differently dated saved baseline.
        np.testing.assert_allclose(yoy.loc[dates], saved.loc[dates, column], atol=1e-8, rtol=1e-8,
                                   err_msg=f"Refresh saved {country} {name} forecasts before building scenarios")
        oil = drivers["oil"].loc[:frame.index[-1]].dropna()
        oil_index = frame.columns.get_loc("oil")
        oil_path = baseline[:, oil_index]
        if not np.isfinite(oil_path).all() or (oil_path <= 0).any():
            raise ValueError(f"Non-positive baseline oil path: {country} {name}")
        drawable = {}
        for driver, metadata in DRAWABLE_DRIVERS.items():
            if driver not in frame:
                continue
            if country == "JP" and driver == "expectations":
                metadata = dict(metadata, label="Tankan expectations", unit="Per cent, 1 year ahead")
            driver_history = drivers[driver].loc[:frame.index[-1]].dropna().iloc[-60:]
            index = frame.columns.get_loc(driver)
            drawable[driver] = dict(metadata, index=index, baseline=baseline[:, index].tolist(),
                                    dates=driver_history.index.to_timestamp().strftime("%Y-%m-%d").tolist(),
                                    history=driver_history.tolist(), last=float(driver_history.iloc[-1]))
        models.append(dict(
            label=label, dates=dates.to_timestamp().strftime("%Y-%m-%d").tolist(), periods=dates.astype(str).tolist(),
            coefficients=fit.coefs.tolist(), intercept=fit.intercept.tolist(),
            seed=frame.to_numpy()[-max(1, fit.k_ar):].tolist(), oilIndex=oil_index,
            oilLast=float(oil.iloc[-1]), oilBaseline=oil_path.tolist(), baseline=baseline.tolist(),
            cpiHistory=history.iloc[-year:].tolist(), year=year, baselineYoy=yoy.loc[dates].tolist(),
            oilDates=oil.iloc[-60:].index.to_timestamp().strftime("%Y-%m-%d").tolist(),
            oilHistory=oil.iloc[-60:].tolist(), lastDate=str(frame.index[-1].to_timestamp().date()),
            lastYoy=float(yoy.loc[frame.index[-1]]), lastPeriod=str(frame.index[-1]), lag=fit.k_ar, drivers=drawable,
        ))
    return models


# =============================================================================
#%% Embed the editor with its data so GitHub Pages needs no model server
def scenario_html(country, spec, figure, include_plotlyjs):
    models = scenario_payload(country, spec)
    key = spec["path"].stem
    if country == "AU":
        figure.add_hrect(y0=2, y1=3, fillcolor="#0B6E4F", opacity=0.10,
                         line_width=0, layer="below",
                         annotation_text="2–3% target", annotation_position="top left",
                         annotation_font=dict(size=10, color="#0B6E4F"))
    elif country in {"JP", "KR"}:
        figure.add_hline(y=2, line_color="#0B6E4F", line_dash="dot", line_width=1.4,
                         annotation_text="2% CPI target", annotation_position="top left",
                         annotation_font=dict(size=10, color="#0B6E4F"))
    figure.update_xaxes(range=[str(pd.Timestamp(models[0]["lastDate"]) - pd.DateOffset(years=3)), models[0]["dates"][-1]],
                        rangeselector=dict(visible=False))
    figure.update_layout(height=420, margin=dict(l=55, r=18, t=70, b=40),
                         legend=dict(font=dict(size=11), x=0, xanchor="left", y=1.03))
    chart = pio.to_html(figure, full_html=False, include_plotlyjs=include_plotlyjs,
                        div_id=f"{key}-inflation", config={"responsive": True, "displaylogo": False})
    unit = "quarter" if spec["frequency"] == "Q" else "month"
    cards = []
    for name, driver in models[0]["drivers"].items():
        cards.append(
            f'<section class="driver-card" data-driver="{name}"><h5>{escape(driver["label"])}</h5>'
            f'<p class="driver-unit">{escape(driver["unit"])}</p>'
            f'<div class="oil-drag-container"><div id="{key}-{name}" class="plotly-graph-div"></div>'
            '<div class="oil-drag-handles"></div></div>'
            '<button type="button" data-role="reset-driver">Reset path</button>'
            '<details class="driver-exact"><summary>Exact values</summary><div class="scenario-controls">'
            '<label>Forecast period <select data-role="period"></select></label>'
            f'<label>{escape(driver["label"])} <input data-role="price" type="number" step="0.01" required></label>'
            '<button type="button" data-role="apply">Apply value</button></div></details></section>'
        )
    availability = ('<p class="scenario-method">Tankan: general-price expectations one year ahead, all enterprises and industries. '
                    'Quarterly readings carry forward from the following month. The estimation sample starts in 2014.</p>'
                    if country == "JP" else "")
    return (
        f'<article class="chart-block oil-scenario" id="{key}" data-country="{country}">'
        f'<h3>{escape(spec["title"])}</h3><p>{escape(spec["description"])}</p>'
        '<div class="scenario-columns"><section class="scenario-forecast"><h4>Inflation forecast</h4>'
        f'{chart}</section><section class="scenario-editor"><h4>Draw driver paths</h4>'
        '<p class="driver-legend"><span class="legend-observed">Observed</span> · '
        '<span class="legend-baseline">Original forecast</span> · <span class="legend-scenario">Drawn scenario</span></p>'
        f'<div class="driver-mini-grid" style="--driver-count:{len(cards)}">{"".join(cards)}</div>'
        f'<p>Draw across the shaded future area of any chart. Each stroke sets the {unit}s it crosses. '
        'All paths stay visible and combine to update the inflation forecasts.</p>'
        '<button type="button" data-role="reset">Reset country forecasts</button>'
        '<p data-role="status" role="status" aria-live="polite"></p>'
        f'{availability}</section></div>'
        '<p class="scenario-method">Coefficients stay fixed. Oil uses price levels; unemployment and expectations use percentages. '
        'Drawn paths apply together to every measure for this country, including companion charts. For each edited driver, '
        'undrawn dates retain each model’s original forecast. Other drivers evolve recursively. Effects enter through lags; '
        'these are mechanical scenarios, not identified causal shocks. Reset restores the original forecasts.</p>'
        f'<script type="application/json" class="scenario-data">{json.dumps(models, allow_nan=False)}</script></article>'
    )

File: Charts/inflation_scenarios.js
// =============================================================================
//%% Fixed-coefficient VAR recursion, shared by the browser and numerical checks
function projectScenario(model, paths = {}) {
  const rows = model.seed.map(row => row.slice());
  const levels = model.cpiHistory.slice();
  const rates = [];
  const yoy = model.dates.map((date, h) => {
    const next = model.intercept.slice();
    model.coefficients.forEach((matrix, lag) => {
      const previous = rows[rows.length - lag - 1];
      matrix.forEach((equation, i) => equation.forEach((weight, j) => { next[i] += weight * previous[j]; }));
    });
    Object.entries(paths).forEach(([name, values]) => {
      const driver = model.drivers[name], value = values[h];
      if (!driver || !Number.isFinite(value) ||
          (driver.minimum !== null && value < driver.minimum) ||
          (driver.maximum !== null && value > driver.maximum)) {
        throw new Error(`Invalid ${name} path value.`);
      }
      next[driver.index] = value;
    });
    if (!next.every(Number.isFinite) || next[0] <= -100) throw new Error('This path produces invalid inflation levels. Try a smaller change.');
    rows.push(next);
    rates.push(next);
    const level = levels[levels.length - 1] * (1 + next[0] / 100);
    const result = 100 * (level / levels[levels.length - model.year] - 1);
    if (!Number.isFinite(result)) throw new Error('This path produces an invalid forecast. Try a smaller change.');
    levels.push(level);
    return result;
  });
  return {yoy, rates};
}

// =============================================================================
//%% Shared driver edits: active paths combine in a single recursion
function projectOilScenario(model, oilPath) {
  return projectScenario(model, {oil: oilPath});
}

function projectCountryScenario(models, edits) {
  return models.map(model => {
    const paths = {};
    edits.forEach((values, name) => {
      if (!values.size || !model.drivers[name]) return;
      paths[name] = model.dates.map((date, i) => values.has(date) ? values.get(date) : model.drivers[name].baseline[i]);
    });
    return projectScenario(model, paths);
  });
}

// =============================================================================
//%% Sample a freehand stroke at forecast dates, including skipped points
function sampleOilStroke(positions, start, end) {
  const nearest = x => positions.reduce((best, value, i) =>
    Math.abs(value - x) < Math.abs(positions[best] - x) ? i : best, 0);
  const first = nearest(start.x), last = nearest(end.x);
  const samples = [];
  for (let i = Math.min(first, last); i <= Math.max(first, last); i++) {
    const fraction = end.x === start.x ? 1 : Math.max(0, Math.min(1, (positions[i] - start.x) / (end.x - start.x)));
    samples.push([i, start.y + fraction * (end.y - start.y)]);
  }
  return samples;
}

// =============================================================================
//%% Freehand driver drawing and synchronized inflation charts
async function initializeOilScenarios() {
  const countries = new Map();
  const editors = Array.from(document.querySelectorAll('.oil-scenario')).map(root => {
    const country = root.dataset.country;
    if (!countries.has(country)) countries.set(country, new Map());
    return {root, country, models: JSON.parse(root.querySelector('.scenario-data').textContent)};
  });
  for (const editor of editors) {
    const {root, models, country} = editor, m = models[0];
    const countryEdits = countries.get(country);
    Object.keys(m.drivers).forEach(name => { if (!countryEdits.has(name)) countryEdits.set(name, new Map()); });
    let active = 'oil';
    const driver = () => m.drivers[active];
    const edits = () => countryEdits.get(active);
    const find = role => root.querySelector(`[data-role="${role}"]`);
    const period = find('period'), price = find('price'), status = find('status'), driverSelect = find('driver');
    const inflation = document.getElementById(`${root.id}-inflation`);
    const oil = document.getElementById(`${root.id}-oil`);
    const overlay = root.querySelector('.oil-drag-handles');
    let currentResults = projectCountryScenario(models, countryEdits);
    const path = () => currentResults[0].rates.map(row => row[driver().index]);
    const clamp = value => Math.max(driver().minimum ?? -Infinity, Math.min(driver().maximum ?? Infinity, value));
    function chartRange() {
      const values = [...driver().history.slice(-m.year * 2), ...path(), ...driver().baseline];
      const low = Math.min(...values), high = Math.max(...values);
      const pad = Math.max((high - low) * 0.5, active === 'oil' ? 30 : 2);
      return [clamp(low - pad), clamp(high + pad)];
    }
    const signed = value => Math.abs(value) < 0.0005 ? '0.000' : `${value > 0 ? '+' : ''}${value.toFixed(3)}`;
    period.replaceChildren(...m.periods.map((label, i) => new Option(`${label} · +${i + 1}`, i)));
    function select(index) {
      period.value = String(index);
      price.value = path()[index].toFixed(2);
    }
    const layout = {
      height: 420, margin: {l: 55, r: 18, t: 70, b: 40},
      font: {family: 'Gill Sans MT, Gill Sans, Arial, sans-serif', color: '#1A1A1A'},
      paper_bgcolor: 'white', plot_bgcolor: 'white', hovermode: 'closest', dragmode: false,
      legend: {orientation: 'h', y: 1.03, yanchor: 'bottom', x: 0, xanchor: 'left', font: {size: 11}},
      xaxis: {type: 'date', showgrid: false, fixedrange: true,
        range: [driver().dates[Math.max(0, driver().dates.length - m.year * 2)], m.dates[m.dates.length - 1]]},
      yaxis: {title: {text: driver().unit}, gridcolor: '#EFEFEF', fixedrange: true,
        range: chartRange()},
    };
    await Plotly.newPlot(oil, [
      {x: driver().dates, y: driver().history, name: 'Observed', mode: 'lines', line: {color: '#0B6E4F'}},
      {x: [m.lastDate, ...m.dates], y: [driver().last, ...driver().baseline], name: 'Original forecast', mode: 'lines', line: {color: '#9A9A9A', dash: 'dash'}},
      {x: m.dates, y: path(), name: 'Scenario path', mode: 'lines', line: {color: '#B5651D', width: 2.5}},
    ], layout, {responsive: true, displaylogo: false, displayModeBar: false});
    const surface = document.createElement('div');
    surface.className = 'oil-draw-surface';
    surface.setAttribute('aria-label', 'Draw the selected driver path across future periods');
    overlay.appendChild(surface);
    const handles = m.dates.map((date, i) => {
      const handle = document.createElement('button');
      handle.type = 'button';
      handle.className = 'oil-handle';
      handle.setAttribute('aria-label', `Driver value for ${m.periods[i]}; use arrow keys to adjust`);
      handle.addEventListener('keydown', event => {
        if (!['ArrowUp', 'ArrowDown'].includes(event.key)) return;
        event.preventDefault();
        select(i);
        change(i, clamp(path()[i] + (event.key === 'ArrowUp' ? driver().keyStep : -driver().keyStep)));
      });
      overlay.appendChild(handle);
      return handle;
    });
    function positionHandles() {
      const {xaxis, yaxis} = oil._fullLayout;
      const left = Math.max(0, xaxis.d2p(m.lastDate));
      Object.assign(surface.style, {left: `${xaxis._offset + left}px`, top: `${yaxis._offset}px`,
        width: `${xaxis._length - left}px`, height: `${yaxis._length}px`});
      path().forEach((value, i) => {
        handles[i].style.left = `${xaxis._offset + xaxis.d2p(m.dates[i])}px`;
        handles[i].style.top = `${yaxis._offset + yaxis.d2p(value)}px`;
        handles[i].title = `${driver().label}, ${m.periods[i]}: ${value.toFixed(2)} ${driver().unit}`;
      });
    }
    oil.on('plotly_afterplot', positionHandles);
    editor.render = function(results) {
      currentResults = results;
      Plotly.restyle(inflation, {
        x: models.map(model => [model.lastDate, ...model.dates]),
        y: models.map((model, i) => [model.lastYoy, ...results[i].yoy]),
      }, models.map((model, i) => 2 * i + 1));
      Plotly.restyle(oil, {y: [path()]}, [2]);
      if (!stroke && path().some(value => value < oil._fullLayout.yaxis.range[0] || value > oil._fullLayout.yaxis.range[1])) {
        Plotly.relayout(oil, {'yaxis.range': chartRange()});
      }
      positionHandles();
      const impacts = models.map((model, i) => {
        const peak = results[i].yoy.reduce((best, value, h) =>
          Math.abs(value - model.baselineYoy[h]) > Math.abs(best) ? value - model.baselineYoy[h] : best, 0);
        return `${model.label}: ${signed(peak)} pp`;
      });
      const activeNames = [...countryEdits].filter(([name, values]) => values.size).map(([name]) => m.drivers[name]?.label ?? name);
      status.textContent = activeNames.length
        ? `Active: ${activeNames.join(", ")}. Largest changes vs original: ${impacts.join('; ')}. Effects start after the model's lag, not in the edited period itself.`
        : 'Original forecasts. Press anywhere in the shaded forecast area and draw across it to update all measures.';
      select(Number(period.value));
    };
    function updateCountry() {
      const peers = editors.filter(item => item.country === country);
      // Validate every result before updating any of the charts.
      const results = peers.map(item => projectCountryScenario(item.models, countryEdits));
      peers.forEach((item, i) => { if (item.render) item.render(results[i]); });
    }
    function change(index, value) {
      const date = m.dates[index], previous = edits().get(date);
      if (!Number.isFinite(value) || value !== clamp(value)) return;
      edits().set(date, value);
      try { updateCountry(); }
      catch (error) {
        if (previous === undefined) edits().delete(date); else edits().set(date, previous);
        status.textContent = error.message;
      }
    }
    let stroke = null;
    function pointerPosition(event) {
      const box = oil.getBoundingClientRect(), {xaxis, yaxis} = oil._fullLayout;
      return {x: Math.max(0, Math.min(xaxis._length, event.clientX - box.left - xaxis._offset)),
        y: Math.max(0, Math.min(yaxis._length, event.clientY - box.top - yaxis._offset))};
    }
    function draw(event) {
      if (!stroke || event.pointerId !== stroke.id) return;
      const end = pointerPosition(event), {xaxis, yaxis} = oil._fullLayout;
      const samples = sampleOilStroke(m.dates.map(date => xaxis.d2p(date)), stroke.point, end);
      const previous = new Map(edits());
      samples.forEach(([i, y]) => edits().set(m.dates[i], clamp(yaxis.p2d(y))));
      try { updateCountry(); }
      catch (error) {
        edits().clear(); previous.forEach((value, date) => edits().set(date, value));
        status.textContent = error.message;
      }
      stroke.point = end;
    }
    surface.addEventListener('pointerdown', event => {
      if (event.button !== 0 || stroke) return;
      event.preventDefault();
      stroke = {id: event.pointerId, point: pointerPosition(event)};
      surface.setPointerCapture(event.pointerId);
      draw(event);
    });
    surface.addEventListener('pointermove', draw);
    surface.addEventListener('pointerup', event => {
      draw(event);
      if (stroke && event.pointerId === stroke.id) stroke = null;
    });
    surface.addEventListener('pointercancel', () => { stroke = null; });
    surface.addEventListener('lostpointercapture', () => { stroke = null; });
    period.addEventListener('change', () => select(Number(period.value)));
    function apply() { if (price.reportValidity()) change(Number(period.value), Number(price.value)); }
    find('apply').addEventListener('click', apply);
    price.addEventListener('change', apply);
    price.addEventListener('keydown', event => { if (event.key === 'Enter') apply(); });
    find('reset').addEventListener('click', () => { countryEdits.forEach(values => values.clear()); updateCountry(); });
    find('reset-driver').addEventListener('click', () => { edits().clear(); updateCountry(); });
    price.required = true;
    driverSelect.addEventListener('change', async () => {
      stroke = null;
      active = driverSelect.value;
      const d = driver();
      find('value-label').textContent = `${d.label} (${d.unit})`;
      if (d.minimum === null) price.removeAttribute('min'); else price.min = d.minimum;
      if (d.maximum === null) price.removeAttribute('max'); else price.max = d.maximum;
      await Plotly.restyle(oil, {x: [d.dates, [m.lastDate, ...m.dates], m.dates],
        y: [d.history, [d.last, ...d.baseline], path()]});
      await Plotly.relayout(oil, {'yaxis.title.text': d.unit, 'yaxis.range': chartRange()});
      positionHandles();
      select(Number(period.value));
    });
    editor.render(projectCountryScenario(models, countryEdits));
  }
}

if (typeof document !== 'undefined') initializeOilScenarios();
if (typeof module !== 'undefined') module.exports = {projectScenario, projectOilScenario, projectCountryScenario, sampleOilStroke};
